gremlin-client/src/io/mod.rs

```diff
@@ -1,34 +1,142 @@
 #[macro_use]
 mod macros;
+mod graph_binary_v1;
 mod serializer_v2;
 mod serializer_v3;
 
 use crate::conversion::ToGValue;
+use crate::message::{ReponseStatus, RequestIdV2, Response, ResponseResult};
 use crate::process::traversal::{Order, Scope};
 use crate::structure::{Cardinality, Direction, GValue, Merge, T};
+use graph_binary_v1::GraphBinaryV1Deser;
+use serde::{Deserialize as SerdeDeserialize, Deserializer};
+use serde_derive::Deserialize;
 use serde_json::{json, Map, Value};
+use std::collections::HashMap;
+use std::convert::TryInto;
 use std::string::ToString;
+use uuid::Uuid;
 
-use crate::{GremlinError, GremlinResult};
+use crate::{io::graph_binary_v1::GraphBinaryV1Ser, GKey, GremlinError, GremlinResult, Message};
 
 #[derive(Debug, Clone)]
-pub enum GraphSON {
-    V2,
-    V3,
+pub enum IoProtocol {
+    GraphSONV2,
+    GraphSONV3,
+    GraphBinaryV1,
 }
 
-impl GraphSON {
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MiddleResponse {
+    pub request_id: Uuid,
+    pub result: MiddleResponseResult,
+    pub status: ReponseStatus,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct MiddleResponseResult {
+    pub data: Value,
+}
+
+impl IoProtocol {
     pub fn read(&self, value: &Value) -> GremlinResult<Option<GValue>> {
         if let Value::Null = value {
             return Ok(None);
         }
         match self {
-            GraphSON::V2 => serializer_v2::deserializer_v2(value).map(Some),
-            GraphSON::V3 => serializer_v3::deserializer_v3(value).map(Some),
+            IoProtocol::GraphSONV2 => serializer_v2::deserializer_v2(value).map(Some),
+            IoProtocol::GraphSONV3 => serializer_v3::deserializer_v3(value).map(Some),
+            IoProtocol::GraphBinaryV1 => todo!(),
+        }
+    }
+
+    pub fn read_response(&self, response: Vec<u8>) -> GremlinResult<Response> {
+        match self {
+            IoProtocol::GraphSONV2 => {
+                let middle_form: MiddleResponse =
+                    serde_json::from_slice(&response).map_err(GremlinError::from)?;
+                Ok(Response {
+                    request_id: Some(middle_form.request_id),
+                    result: ResponseResult {
+                        data: serializer_v2::deserializer_v2(&middle_form.result.data).map(Some)?,
```

Comment:
This is a notable change. Previously the `result.data` would only be parsed depending on the status code. To normalize the disjoint (de)serializers I go ahead and attempt to parse it regardless, but handle its omission via a `None` in the normalized type.